core/src/apps/cardano/address.py

```diff
@@ -43,19 +46,83 @@ def validate_address_parameters(parameters: CardanoAddressParametersType) -> Non
+                parameters.address_n_staking, parameters.staking_key_hash
+            )
+
+        elif parameters.address_type == CardanoAddressType.BASE_SCRIPT_KEY:
+            _validate_base_address_stake_credential(
                 parameters.address_n_staking, parameters.staking_key_hash
             )
+            if (
+                not parameters.script_payment_hash
+                or len(parameters.script_payment_hash) != SCRIPT_HASH_SIZE
+            ):
+                raise INVALID_ADDRESS_PARAMETERS
+
+        elif parameters.address_type == CardanoAddressType.BASE_KEY_SCRIPT:
+            if not is_shelley_path(parameters.address_n):
+                raise INVALID_ADDRESS_PARAMETERS
+            if (
+                not parameters.script_staking_hash
+                or len(parameters.script_staking_hash) != SCRIPT_HASH_SIZE
+            ):
+                raise INVALID_ADDRESS_PARAMETERS
+
+        elif parameters.address_type == CardanoAddressType.BASE_SCRIPT_SCRIPT:
+            if (
+                not parameters.script_payment_hash
+                or len(parameters.script_payment_hash) != SCRIPT_HASH_SIZE
+            ):
+                raise INVALID_ADDRESS_PARAMETERS
+            if (
+                not parameters.script_staking_hash
+                or len(parameters.script_staking_hash) != SCRIPT_HASH_SIZE
+            ):
+                raise INVALID_ADDRESS_PARAMETERS
+
         elif parameters.address_type == CardanoAddressType.POINTER:
+            if not is_shelley_path(parameters.address_n):
+                raise INVALID_ADDRESS_PARAMETERS
             if parameters.certificate_pointer is None:
                 raise INVALID_ADDRESS_PARAMETERS
+
+        elif parameters.address_type == CardanoAddressType.POINTER_SCRIPT:
+            if (
+                not parameters.script_payment_hash
+                or len(parameters.script_payment_hash) != SCRIPT_HASH_SIZE
+            ):
+                raise INVALID_ADDRESS_PARAMETERS
+            if parameters.certificate_pointer is None:
+                raise INVALID_ADDRESS_PARAMETERS
+
+        elif parameters.address_type == CardanoAddressType.ENTERPRISE:
+            if not is_shelley_path(parameters.address_n):
+                raise INVALID_ADDRESS_PARAMETERS
+
+        elif parameters.address_type == CardanoAddressType.ENTERPRISE_SCRIPT:
+            if (
+                not parameters.script_payment_hash
+                or len(parameters.script_payment_hash) != SCRIPT_HASH_SIZE
+            ):
+                raise INVALID_ADDRESS_PARAMETERS
+
         elif parameters.address_type == CardanoAddressType.REWARD:
-            if not SCHEMA_STAKING_ANY_ACCOUNT.match(parameters.address_n):
+            if not is_shelley_path(parameters.address_n_staking):
+                raise INVALID_ADDRESS_PARAMETERS
+            if not SCHEMA_STAKING_ANY_ACCOUNT.match(parameters.address_n_staking):
+                raise INVALID_ADDRESS_PARAMETERS
```

Comment:
The rest of the path is validate against a schema later in `get_payment_credential_policy` and the path is then shown to the user with a warning if it's unusual. Perhaps I could just add a comment here that that's happening.

Reply:
I can't come up with an appropriate comment to place here so I'm wandering if the comment is actually needed since this is proper validation for an address path, no? If the user agrees it's ok to derive an address with this path, no? The only required thing is that it is a shelley path. Or am I wrong?